### PR TITLE
Feature/plotting

### PR DESCRIPTION
**Pull request by albertodepaola**
# What does this PR do?
Adding support to save fine tuning metrics to json. Adding script to plot saved metrics.
Tracks loss and perplexity, both at the step and epoch level and saves them to a json file. Tracks both train and validation metrics, for comparison after finetuning.

This requires matplotlib to be added as a requirement

## Feature/Issue validation/testing

- [x] Run finetuning on a single GPU with the default dataset to verify everything is working correctly

- [x] Run finetuning on multiple GPUs with default dataset to verify it continues working

1. Running tests:
image

2. Command on a fully configured environment: 
```
torchrun --nnodes 1 --nproc_per_node 4  examples/finetuning.py --enable_fsdp --use_peft --peft_method lora --mode
l_name ../Llama-2-7b-chat-hf/ --pure_bf16 --output_dir ../Llama-2-7b-peft-samsum-10 --batch_size_training 2 --gradient_accumulation_steps 2 --num_epochs 15 --save_metrics
```

3. Results:
metrics_data_3-2023-09-22_03-11-30_train_and_validation_loss
metrics_data_3-2023-09-22_03-11-30_train_and_validation_perplexity

A test was added to verify that when the save_metrics is used in train_config, the file gets generated.

### CODE REVIEW

**Comment by albertodepaola** [comment]
Executing tests with full datasets to attach to the PR, but wanted to start the conversation on the proposed changes.

**Review by HamidShojanazeri — COMMENTED** [review]
thanks @albertodepaola for the PR, this is a great feature. Couple of questions, was wondering if tensorboard path has been explored as well?

Also, it would be very helpful to:

a- make this function configurable/ if user choose to do so.
b - please add a flag and add explanation in the readme.

**Comment by albertodepaola** [comment]
Added the `--save_metrics` parameter to configure if loss and perplexity are saved to a json file. Added comment in the readme and unit test to the corresponding file.

**Review by jeffxtang — APPROVED** [review]
LGTM
